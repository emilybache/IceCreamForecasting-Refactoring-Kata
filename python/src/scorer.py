import enum
import random
import requests


class IceCream(enum.Enum):
    Strawberry = 1
    Chocolate = 2
    Vanilla = 3


flavour = None


def get_score():
    sunny_today = lookup_weather()
    return get_score_with_weather_and_flavour(sunny_today, flavour)


def get_score_with_weather_and_flavour(sunny_today, current_flavour=None):
    if current_flavour == IceCream.Strawberry:
        if sunny_today:
            return 10
        else:
            return 5
    elif current_flavour == IceCream.Chocolate:
        return 6
    elif current_flavour == IceCream.Vanilla:
        if sunny_today:
            return 7
        else:
            return 5
    else:
        return -1


def lookup_weather(location=None, days_forward=None):
    location = location or (59.3293, 18.0686)  # default to Stockholm
    days_forward = days_forward or 0
    params = {"latitude": location[0], "longitude": location[1], "days_forward": days_forward}
    weather_app = "http://127.0.0.1:3005"
    try:
        response = requests.get(weather_app + "/forecast", params=params)
    except requests.exceptions.ConnectionError as err:
        raise RuntimeError("Weather service unavailable")
    if response.status_code != 200:
        raise RuntimeError("Weather service request failed")
    forecast = response.json()
    return bool(forecast["weather"]["main"] == "Sunny")


def update_selection():
    score = get_score()
    if score > 5:
        global flavour
        # placeholder implementation - real version would use machine learning to predict sales
        flavour = random.choice([IceCream.Strawberry, IceCream.Chocolate, IceCream.Vanilla])


def get_sales_forecast():
    forecasts = {
        IceCream.Strawberry: 9,
        IceCream.Vanilla: 11,
        IceCream.Chocolate: 3
    }
    return forecasts[flavour]
