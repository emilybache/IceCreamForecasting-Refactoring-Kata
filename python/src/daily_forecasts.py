import datetime

import scorer
from scorer import IceCream


def print_sales_forecasts(now=None):
    def update_selection():
        scorer.update_selection()

    print_sales_forecase_and_update_selection(now, update_selection)


def print_sales_forecase_and_update_selection(now, update_selection):
    names = ["Steve", "Julie", "Francis"]
    now = now or datetime.datetime.now()
    print(f"Forecast at time {now}")
    for name in names:
        if name == "Steve":
            scorer.flavour = IceCream.Strawberry
        else:
            update_selection()
        score = scorer.get_sales_forecast()
        print(f"{name} score: {score}")
