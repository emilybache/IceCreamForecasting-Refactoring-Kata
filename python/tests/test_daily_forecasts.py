import datetime
import approvaltests

from daily_forecasts import print_sales_forecasts, print_sales_forecase_and_update_selection
from scorer import IceCream
import scorer


def stub_update_selection():
    scorer.flavour = IceCream.Vanilla


def test_sales_forecasts(capsys):
    print_sales_forecase_and_update_selection(
        now=datetime.datetime.fromisoformat("2023-05-08 15:26:03.797869"),
        update_selection=stub_update_selection
    )
    output = capsys.readouterr()
    approvaltests.verify(output.out)
