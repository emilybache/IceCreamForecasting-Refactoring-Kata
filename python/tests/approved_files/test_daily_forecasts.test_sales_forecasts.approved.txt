Forecast at time 2023-05-08 15:26:03.797869
Steve score: 9
Julie score: 11
Francis score: 11
